dojo/finding_group/redis.py

```diff
@@ -0,0 +1,248 @@
+            group_ids = json.loads(group_ids_raw) if group_ids_raw else []
+            if finding_group_id not in group_ids:
+                group_ids.append(finding_group_id)
+            redis_client.hset(id_map_key, finding.id, json.dumps(group_ids))
+
+    # This method is used in finding_groups table to show SLA
+    def get_days_remaining(self) -> str:
+        if self.sla_finding_id:
+            finding = Finding.objects.filter(id=self.sla_finding_id).first()
+            days_remaining = finding.sla_days_remaining()
+            severity = finding.severity
+            sla_start_date = finding.get_sla_start_date().strftime("%b %d, %Y")
+            status = "age-green"
+            status_text = f"Remediation for {severity.lower()} findings due in {days_remaining} days or less (started {sla_start_date})"
+            if days_remaining and days_remaining < 0:
+                status = "age-red"
+                status_text = f"Overdue: Remediation for {severity.lower()} findings overdue {days_remaining} days (started {sla_start_date})"
+                days_remaining = abs(days_remaining)
+        elif any(
+            Finding.objects.filter(
+                id__in=self.finding_ids,
+                active=True,
+            ),
+        ):
+            status = "severity-Info"
+            status_text = "No SLA set, but at least one finding is active"
+            days_remaining = "No SLA"
+        else:
+            status = "age-blue"
+            status_text = "No active finding"
+            days_remaining = "Concluded"
+        title = (
+            f'<a class="has-popover" data-toggle="tooltip" data-placement="bottom" title="" href="#" data-content="{status_text}">'
+            f'<span class="label severity {status}">{days_remaining}</span></a>'
+        )
+        return mark_safe(title)
+
+
+@lru_cache(maxsize=1)
+def get_redis_client() -> redis.Redis:
+    return redis.Redis(host="redis", port=6379, decode_responses=True)
+
+
+def get_user_mode(user_id: int) -> GroupMode | None:
+    redis_client = get_redis_client()
+    value = redis_client.hget(USER_MODES_KEY, str(user_id))
+    if value and value not in [m.value for m in GroupMode]:
+        logger.warning(f"Invalid group mode '{value}' found in Redis for user {user_id}, resetting to None.")
+        redis_client.hdel(USER_MODES_KEY, str(user_id))
+        return None
+    return GroupMode(value) if value else None
+
+
+def set_user_mode(user_id: int, mode: GroupMode) -> None:
+    redis_client = get_redis_client()
+    redis_client.hset(USER_MODES_KEY, str(user_id), mode.value)
+    logger.info(f"User {user_id} dynamic finding groups mode set to {mode.value}")
+
+
+def load_or_rebuild_finding_groups(mode: GroupMode) -> dict[str, DynamicFindingGroups]:
+    redis_client = get_redis_client()
+    fg_key = DynamicFindingGroups.get_fg_key(mode)
+    id_map_key = DynamicFindingGroups.get_id_map_key(mode)
+
+    if not redis_client.exists(SYSTEM_CHANGE):
```

Comment:
We should write some comments here or a README explaining the management of the information on Redis. In particular, here we are managing a global timestamp that we use to detect if updating the previously-stored groups is necessary.